File: Data Structures & Algorithms/valid-binary-search-tree/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution
{
public:
    bool isValidBST(TreeNode *root)
    {
        return checkBST(root, numeric_limits<int>::min(), numeric_limits<int>::max());
    }

private:
    bool checkBST(TreeNode *root, int bot, int top)
    {
        if (root == nullptr)
        {
            return true;
        }
        if (root->val <= bot || root->val >= top)
        {
            return false;
        }
        return checkBST(root->left, bot, root->val) && checkBST(root->right, root->val, top);
    }
};

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        if not root:
            return 0
        
        mp = {root: root.val}
        queue = deque([root])

        res = 0
        while queue:
            node = queue.popleft()
            if node.val >= mp[node]:
                res += 1
            
            if node.left:
                mp[node.left] = max(mp[node], node.left.val)
                queue.append(node.left)
            if node.right:
                mp[node.right] = max(mp[node], node.right.val)
                queue.append(node.right)
        return res